Implement CRI stats in Docker Shim

bobbypage opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
This PR implements CRI Stats in the Docker Shim. It is needed to enable CRI stats for Docker and ongoing /stats/summary API changes in moving to use CRI.

Related issues:
#46984 (CRI: instruct kubelet to (optionally) consume container stats from CRI)
#45614 (CRI: add methods for container stats) 

This PR is also a followup to my original PR (https://github.com/kubernetes/kubernetes/pull/50396) to implement Windows Container Stats. The plan is that Windows Stats will use a hybrid model: pod and container level stats will come from CRI (via dockershim) and that node level stats will come from a "winstats" package that exports cadvisor like datastructures using windows specific perf counters from the node. I will update that PR to only export node level stats. 

@yujuhong @yguo0905 @dchen1107 @jdumars @anhowe @michmike

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
```

bobbypage committed:
Implement CRI stats in dockershim for Windows

Implement CRI stats for dockershim using docker stats. This enables use
of the summary api to get container metrics on Windows where CRI stats
are enabled.

bobbypage commented:
/cc @yujuhong @yguo0905

feiskyer commented:
This PR calls docker's inspect API for each container stats request, I'm worried about introducing huge loads to dockerd.

#46984 is working on refactor cadvisor to implement the container stats, I think it's better to call cadvisor for linux (after #46984) and only get from docker's API for windows.

jdumars commented:
/ok-to-test

bobbypage commented:
/test pull-kubernetes-e2e-gce-etcd3

yujuhong commented:
> #46984 is working on refactor cadvisor to implement the container stats, I think it's better to call cadvisor for linux (after #46984) and only get from docker's API for windows.

We'll still use cadvisor with for docker containers on linux. Kubelet needs to be able to choose cadvisor if it is on linux.

The question remains is whether this PR should implement the container metrics for both linux AND windows. Given that we have no plans to actually use this on linux, I'd suggest move all the code to `_windows.go` file and leave those functions on linux as "not implemented".

@feiskyer @yguo0905 thoughts?

yguo0905 commented:
> Given that we have no plans to actually use this on linux, I'd suggest move all the code to _windows.go file and leave those functions on linux as "not implemented".

Yes, makes sense.

yguo0905 commented on pkg/kubelet/dockershim/docker_stats.go:
```diff
@@ -17,16 +17,112 @@ limitations under the License.
+		},
+		WritableLayer: &runtimeapi.FilesystemUsage{
+			Timestamp: time.Now().UnixNano(),
+			StorageId: &runtimeapi.StorageIdentifier{Uuid: containerJSON.Driver},
```
`containerJSON.Driver` is not the correct info we want. This is the UUID of the filesystem, which is used by kubelet to get the filesystem stats from cadvisor. It's not supported on Windows. I'm not sure we should change cadvisor to support something like this on Windows or probably change CRI to return the stats instead of the UUID in this case.

@yujuhong

feiskyer replied:
> This is the UUID of the filesystem, which is used by kubelet to get the filesystem stats from cadvisor. It's not supported on Windows. 

Same questions here: does windows support filesystem UUID? or does all fields (e.g. InodesUsed) in stats API make sense to windows?

yujuhong replied:
@bobbypage, could you briefly explain how these fields are mapped to concepts in windows? Thanks.

bobbypage replied:
as @yguo0905 mentioned, I don't believe Windows has a UUID of the file system, so there is no equivalent for `StorageId`. Additionally, some of the fields under `FilesystemUsage`, like [`InodesUsed`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/v1alpha1/runtime/api.pb.go#L3003) don't have corresponding fields on Windows. I will update the PR and remove the `StorageId` usage here.

feiskyer commented:
> The question remains is whether this PR should implement the container metrics for both linux AND windows. Given that we have no plans to actually use this on linux, I'd suggest move all the code to _windows.go file and leave those functions on linux as "not implemented".

looks weird of same api implemented in different packages for windows and linux. I think it's better to impl this in dockershim for both platforms.

bobbypage commented:
@yujuhong @yguo0905 I have moved the code to _windows.go files as you suggested.

yujuhong commented:
> The question remains is whether this PR should implement the container metrics for both linux AND windows. Given that we have no plans to actually use this on linux, I'd suggest move all the code to _windows.go file and leave those functions on linux as "not implemented".

@bobbypage I was mainly talking about files such as `dockershim/docker_stats.go`.

yujuhong commented on pkg/kubelet/dockershim/docker_stats.go:
```diff
@@ -17,16 +17,112 @@ limitations under the License.
+}
+
+func (ds *dockerService) getMemory(dockerStats *dockertypes.Stats) uint64 {
+	if runtime.GOOS == "windows" {
```
This file should be split into _windows.go and _linux.go, so the if/else will not be needed.

bobbypage replied:
Fixed.

yujuhong commented on pkg/kubelet/dockershim/libdocker/kube_docker_client.go:
```diff
@@ -123,6 +123,19 @@ func (d *kubeDockerClient) InspectContainer(id string) (*dockertypes.ContainerJS
+func (d *kubeDockerClient) InspectContainerWithSize(id string) (*dockertypes.ContainerJSON, error) {
+	ctx, cancel := d.getTimeoutContext()
+	defer cancel()
+	containerJSON, _, err := d.client.ContainerInspectWithRaw(ctx, id, true)
```
Add comments explaining what this call does -- `Return the size of container as fields SizeRw and SizeRootFs.`

bobbypage replied:
Fixed, thanks for suggestion.

bobbypage commented:
@yujuhong can you clarify what files you believe should have a _windows.go counterpart? Is it just `pkg/kubelet/dockershim/docker_stats.go` or is it `pkg/kubelet/dockershim/docker_stats.go`, `pkg/kubelet/remote/remote_runtime_windows.go`, and `pkg/kubelet/dockershim/remote/docker_service_windows.go`?

yujuhong commented on pkg/kubelet/dockershim/docker_stats.go:
```diff
@@ -17,16 +17,112 @@ limitations under the License.
 package dockershim
```
Split the file into _linux and windows.go

bobbypage replied:
I have split the file as you suggested.

yujuhong commented on pkg/kubelet/dockershim/remote/docker_service_linux.go:
```diff
@@ -0,0 +1,33 @@
+// +build linux
```
Don't think we need to split this file.

bobbypage replied:
Fixed.

yujuhong commented on pkg/kubelet/remote/remote_runtime_linux.go:
```diff
@@ -0,0 +1,34 @@
+// +build linux
```
No need to split this file as well.

bobbypage replied:
Fixed.

bobbypage commented:
/retest

bobbypage commented:
/test pull-kubernetes-cross

yguo0905 commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+			Annotations: status.Annotations,
+		},
+		Cpu: &runtimeapi.CpuUsage{
+			Timestamp:            time.Now().UnixNano(),
```
Calling `time.Now().UnixNano()` once and use it for all timestamps should be good.

bobbypage replied:
Fixed.

yguo0905 commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+// ContainerStats returns stats for a container stats request based on container id.
+func (ds *dockerService) ContainerStats(containerStatsRequest *runtimeapi.ContainerStatsRequest) (*runtimeapi.ContainerStatsResponse, error) {
+	containerStats, err := ds.createContainerStats(containerStatsRequest.ContainerId)
+
```
The empty line can be removed (and other places as well).

bobbypage replied:
Fixed.

yguo0905 commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+	return containerStats, nil
+}
+
+func (ds *dockerService) getCPU(dockerStats *dockertypes.Stats) uint64 {
```
nit: doesn't need to be a "member function".

bobbypage replied:
Fixed. I deleted this function and moved it inline as per https://github.com/kubernetes/kubernetes/pull/51152#discussion_r139030923

yguo0905 reviewed (commented):
Have we tested this manually?

yujuhong commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+			Annotations: status.Annotations,
+		},
+		Cpu: &runtimeapi.CpuUsage{
+			Timestamp:            time.Now().UnixNano(),
```
Get the timestamp once at line 91. No need to call it multiple times when populating the stats.

bobbypage replied:
Fixed.

yujuhong commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+	}, nil
+}
+
+func (ds *dockerService) createContainerStats(containerID string) (*runtimeapi.ContainerStats, error) {
```
I think `getContainerStats` is closer to what this function does.

bobbypage replied:
Fixed, renamed it to `getContainerStats`

yujuhong commented on pkg/kubelet/dockershim/libdocker/fake_client.go:
```diff
@@ -854,3 +871,10 @@ func (f *FakeDockerPuller) GetImageRef(image string) (string, error) {
+	f.Lock()
+	defer f.Unlock()
+	f.appendCalled(calledDetail{name: "getContainerStats"})
+	return nil, nil
```
`return nil, fmt.Errorf("not implemented")` to make it clear that this function is not properly faked. If anyone wants to use this in a test in the future, they'll need to make sure a valid output is returned.

bobbypage replied:
Fixed.

yujuhong commented on pkg/kubelet/dockershim/libdocker/kube_docker_client.go:
```diff
@@ -522,6 +536,28 @@ func (d *kubeDockerClient) ResizeContainerTTY(id string, height, width uint) err
+	defer cancel()
+
+	response, err := d.client.ContainerStats(ctx, id, false)
+
```
int: remove the line

bobbypage replied:
Fixed.

yujuhong commented on pkg/kubelet/dockershim/libdocker/kube_docker_client.go:
```diff
@@ -123,6 +123,20 @@ func (d *kubeDockerClient) InspectContainer(id string) (*dockertypes.ContainerJS
 	return &containerJSON, nil
 }
 
+func (d *kubeDockerClient) InspectContainerWithSize(id string) (*dockertypes.ContainerJSON, error) {
```
nit: Add a comment that this is only used on windows.

yujuhong commented on pkg/kubelet/dockershim/libdocker/kube_docker_client.go:
```diff
@@ -522,6 +536,28 @@ func (d *kubeDockerClient) ResizeContainerTTY(id string, height, width uint) err
 	})
 }
 
+func (d *kubeDockerClient) GetContainerStats(id string) (*dockertypes.StatsJSON, error) {
```
nit: Add a comment that this is only used on windows.

bobbypage replied:
Fixed.

yujuhong commented on pkg/kubelet/dockershim/libdocker/kube_docker_client.go:
```diff
@@ -522,6 +536,28 @@ func (d *kubeDockerClient) ResizeContainerTTY(id string, height, width uint) err
+	dec := json.NewDecoder(response.Body)
+	var stats dockertypes.StatsJSON
+	err = dec.Decode(&stats)
+
```
int: remove the line

bobbypage replied:
Fixed.

yujuhong commented:
> Have we tested this manually?

+1. Testing and seeing the output of /stats/summary would be great.

bobbypage commented on pkg/kubelet/dockershim/docker_stats_unsupported.go:
```diff
@@ -0,0 +1,34 @@
+package dockershim
+
+import (
+	"fmt"
```
Fixed, thanks for info on the style.

bobbypage commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+
+import (
+	dockertypes "github.com/docker/docker/api/types"
+	runtimeapi "k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime"
```
Fixed.

bobbypage commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+// ContainerStats returns stats for a container stats request based on container id.
+func (ds *dockerService) ContainerStats(containerStatsRequest *runtimeapi.ContainerStatsRequest) (*runtimeapi.ContainerStatsResponse, error) {
+	containerStats, err := ds.createContainerStats(containerStatsRequest.ContainerId)
+
```
Fixed.

bobbypage commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+	var stats []*runtimeapi.ContainerStats
+	for _, container := range containers {
+		containerStats, err := ds.createContainerStats(container.Id)
+
```
Fixed.

bobbypage commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+
+func (ds *dockerService) createContainerStats(containerID string) (*runtimeapi.ContainerStats, error) {
+	statsJSON, err := ds.client.GetContainerStats(containerID)
+
```
Fixed.

bobbypage commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+	}
+
+	containerJSON, err := ds.client.InspectContainerWithSize(containerID)
+
```
Fixed.

bobbypage commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+	}
+
+	status, err := ds.ContainerStatus(containerID)
+
```
Fixed.

bobbypage commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+	return containerStats, nil
+}
+
+func (ds *dockerService) getCPU(dockerStats *dockertypes.Stats) uint64 {
```
Fixed, folded this into `createContainerStats()`

bobbypage commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+
+func (ds *dockerService) getCPU(dockerStats *dockertypes.Stats) uint64 {
+	// have to multiply cpu usage by 100 since docker stats units is in 100's of nano seconds for Windows
+	// see https://github.com/moby/moby/blob/master/api/types/stats.go#L22
```
Thanks for the tip, fixed.

bobbypage commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,119 @@
+	}
+
+	containers, err := ds.ListContainers(filter)
+
```
Fixed.

bobbypage commented:
I have tested this manually on Windows. I had to make a few changes to get it to work correctly.

I tested this manually be merging my other PR for Windows stats (#50396) as well as @yujuhong PR to enable CRI metrics (#52451). 

I had to make an additional change regarding ImageFS to get the summary api to work correctly. @yguo0905 changed the summary api to fail if [ImageFSStats fails](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/stats/summary.go#L63)
 (which it does on Windows).

Since Windows does not support ImageFS, there were two issues:
1. In ` pkg/kubelet/stats/cri_stats_provider.go` there was a nil pointer deference of `storageId`. I have fixed this in [this commit](https://github.com/kubernetes/kubernetes/pull/51152/commits/2ce49d4123cf3b5e19ce66443eb6418047465846).
2. In `pkg/kubelet/stats/cri_stats_provider.go` at least [one image file system is expected](https://github.com/bobbypage/kubernetes/blob/d1e5f784690e1b21291dbf5f23ccd774e7751bd1/pkg/kubelet/stats/cri_stats_provider.go#L151-L172). To work around this, I added a fake filesystem that windows reports in [this commit](https://github.com/kubernetes/kubernetes/pull/51152/commits/9bc873f76bbd8e89ff86765397a6f4480e4d0635).

With these changes the summary api is working as expected with pod stats:
![Screenshot 1](https://i.imgur.com/VgjhISo.png)
![Screenshot 2](https://i.imgur.com/iU5OMx7.png)

yguo0905 commented:
> I had to make an additional change regarding ImageFS to get the summary api to work correctly. @yguo0905 changed the summary api to fail if ImageFSStats fails
(which it does on Windows).

> In pkg/kubelet/stats/cri_stats_provider.go at least one image file system is expected. To work around this, I added a fake filesystem that windows dockershim reports in this commit.

It's always been requiring a full response. We'd like to support best effort partial response https://github.com/kubernetes/heapster/issues/1793.

> In `pkg/kubelet/stats/cri_stats_provider.go` there was a nil pointer deference of `storageId`. I have fixed this in this [commit](https://github.com/kubernetes/kubernetes/pull/51152/commits/2ce49d4123cf3b5e19ce66443eb6418047465846).

Thanks for the fix! I think we might want this fix as a separate PR which can be patched to 1.8. @yujuhong, WDYT?

BTW, we also need change line 261 to `storageID := stats.GetWritableLayer().GetStorageId()` so that it can handle a nil `WritableLayer`.

bobbypage commented:
@yguo0905 Great, I added a [fix](https://github.com/kubernetes/kubernetes/pull/51152/commits/83b54e2707d5f9e9afc4dc7f00ca826fa8fb6659) to handle a possible nil WritableLayer as you mentioned.

bobbypage commented:
/retest

yujuhong commented:
> Thanks for the fix! I think we might want this fix as a separate PR which can be patched to 1.8. @yujuhong, WDYT?

Yep. @bobbypage could you create a separate PR with the fixes and cc both us? If you don't have time, let us know and one of us can do it. Thanks!

bobbypage commented:
Sure, I created a PR [here](https://github.com/kubernetes/kubernetes/pull/52588) with those two fixes.

Should I remove these fixes in this PR since they will be handled by that separate PR?

yujuhong commented:
> Should I remove these fixes in this PR since they will be handled by that separate PR?

Yep. Thanks for the fixes!

The PR looks good to me overall, but I'd like to leave @yguo0905 a chance to lgtm
Please also squash the commits.

/approve

bobbypage commented:
Thanks! I've removed those two fixes from this PR and squashed.

yguo0905 commented on pkg/kubelet/dockershim/docker_stats_windows.go:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	dockerStats := statsJSON.Stats
+	timeStamp := time.Now().UnixNano()
```
Let's call this `timestamp`.

yguo0905 commented on pkg/kubelet/dockershim/libdocker/instrumented_client.go:
```diff
@@ -251,3 +260,11 @@ func (in instrumentedInterface) ResizeContainerTTY(id string, height, width uint
 	recordError(operation, err)
 	return err
 }
+func (in instrumentedInterface) GetContainerStats(id string) (*dockertypes.StatsJSON, error) {
```
nit: add an empty line.

yguo0905 commented on pkg/kubelet/dockershim/docker_image_linux.go:
```diff
@@ -1,5 +1,7 @@
+
 /*
-Copyright 2017 The Kubernetes Authors.
+Copyright 2016 The Kubernetes Authors.
```
2017?

yguo0905 commented on pkg/kubelet/dockershim/docker_image_windows.go:
```diff
@@ -0,0 +1,40 @@
+// ImageFsInfo returns information of the filesystem that is used to store images.
+func (ds *dockerService) ImageFsInfo(req *runtimeapi.ImageFsInfoRequest) (*runtimeapi.ImageFsInfoResponse, error) {
+	// For Windows Stats to work correctly, a file system must be provided. For now, provide a fake filesystem.
+	var filesystems []*runtimeapi.FilesystemUsage
```
nit:

```
filesystems := []*runtimeapi.FilesystemUsage{
    {
        Timestamp:  time.Now().UnixNano(),
        UsedBytes:  &runtimeapi.UInt64Value{Value: 0},
        InodesUsed: &runtimeapi.UInt64Value{Value: 0},
    }
}
```

yguo0905 reviewed (commented):
Looks good. Will LGTM after the small nits are addressed.

bobbypage commented:
Thanks @yguo0905. I have fixed the nits you mentioned.

yguo0905 commented:
/lgtm
Thanks for the PR!

yguo0905 commented:
FYI, https://github.com/kubernetes/kubernetes/pull/52635 is a fix expected to be merged in 1.8. This PR will probably need to be rebased after that, but the conflict should be small.

bobbypage commented:
/retest

bobbypage commented:
I have updated this PR to fix conflicts with #52635. @yujuhong and @yguo0905 can you please take a look and LGTM again? Thanks!

yujuhong commented:
lgtm.
@bobbypage squash the two commits?

bobbypage commented:
I have squashed again.

yguo0905 commented:
/lgtm
